File: integrations/php/oicana-php/tests/ExportOnceTest.php
<?php

declare(strict_types=1);

use Oicana\CompilationMode;
use Oicana\Configuration;
use Oicana\DiagnosticColor;
use Oicana\ExportFormat;
use Oicana\ExportOnceResult;
use Oicana\Template;
use Oicana\ZipLimits;

beforeEach(function () {
    if (!extension_loaded('oicana')) {
        test()->markTestSkipped('oicana extension not loaded');
    }

    if (!file_exists(e2e_template_path())) {
        test()->markTestSkipped('E2E template not found. Run `oicana pack` in e2e-tests/template first.');
    }
});

function pack_minimal_template(string $mainTypst): string
{
    $manifest = <<<TOML
        [package]
        name = "export-once-test"
        version = "0.1.0"
        entrypoint = "main.typ"

        [tool.oicana]
        manifest_version = 1
        TOML;

    $path = tempnam(sys_get_temp_dir(), 'oicana-test-') . '.zip';
    $zip = new ZipArchive();
    $zip->open($path, ZipArchive::CREATE | ZipArchive::OVERWRITE);
    $zip->addFromString('typst.toml', $manifest);
    $zip->addFromString('main.typ', $mainTypst);
    $zip->setCompressionName('typst.toml', ZipArchive::CM_STORE);
    $zip->setCompressionName('main.typ', ZipArchive::CM_STORE);
    $zip->close();

    $bytes = file_get_contents($path);
    assert(is_string($bytes));
    unlink($path);
    return $bytes;
}

test('exportOnce exports without warnings', function () {
    $templateBytes = file_get_contents(e2e_template_path());

    $result = Template::exportOnce($templateBytes, mode: CompilationMode::Development);

    expect($result)->toBeInstanceOf(ExportOnceResult::class)
        ->and($result->document)->toStartWith('%PDF')
        ->and($result->warnings)->toBeNull();
});

test('exportOnce surfaces warnings', function () {
    $templateBytes = pack_minimal_template(
        "#set text(font: \"NonexistentFontExportOnce\")\nContent"
    );

    $result = Template::exportOnce(
        $templateBytes,
        exportFormat: ExportFormat::svg(),
        mode: CompilationMode::Development
    );

    expect($result->document)->toContain('<svg')
        ->and($result->warnings)->not->toBeNull()
        ->and($result->warnings)->toContain('NonexistentFontExportOnce');
});

test('exportOnce enforces zip limits', function () {
    $templateBytes = file_get_contents(e2e_template_path());

    expect(fn() => Template::exportOnce(
        $templateBytes,
        mode: CompilationMode::Development,
        limits: new ZipLimits(maxEntries: 1)
    ))->toThrow(Exception::class, 'entries');
});

test('registration enforces zip limits', function () {
    $templateBytes = file_get_contents(e2e_template_path());

    expect(fn() => new Template(
        $templateBytes,
        limits: new ZipLimits(maxEntries: 1)
    ))->toThrow(Exception::class, 'entries');
});

test('diagnostic color configuration succeeds', function () {
    Configuration::setDiagnosticColor(DiagnosticColor::Ansi);
    Configuration::setDiagnosticColor(DiagnosticColor::None);

    expect(true)->toBeTrue();
});

File: integrations/java/oicana/src/test/java/com/oicana/ExportOnceTest.java
package com.oicana;

import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for the one-shot {@link Template#exportOnce} API.
 */
class ExportOnceTest {

    private static final Path PROJECT_ROOT = Path.of("").toAbsolutePath()
            .getParent().getParent().getParent();

    private static final String MINIMAL_MANIFEST = """
            [package]
            name = "export-once-test"
            version = "0.1.0"
            entrypoint = "main.typ"

            [tool.oicana]
            manifest_version = 1
            """;

    private static byte[] templateFile() throws IOException {
        return Files.readAllBytes(
                PROJECT_ROOT.resolve("e2e-tests/template/oicana-e2e-test-x.y.z.zip"));
    }

    private static byte[] packTemplate(String manifest, String mainTypst) throws IOException {
        var stream = new ByteArrayOutputStream();
        try (var zip = new ZipOutputStream(stream)) {
            zip.setMethod(ZipOutputStream.STORED);
            for (var entry : Map.of("typst.toml", manifest, "main.typ", mainTypst).entrySet()) {
                byte[] content = entry.getValue().getBytes(StandardCharsets.UTF_8);
                var zipEntry = new ZipEntry(entry.getKey());
                zipEntry.setSize(content.length);
                var crc = new java.util.zip.CRC32();
                crc.update(content);
                zipEntry.setCrc(crc.getValue());
                zip.putNextEntry(zipEntry);
                zip.write(content);
                zip.closeEntry();
            }
        }
        return stream.toByteArray();
    }

    @Test
    void exportsWithoutWarnings() throws IOException {
        ExportOnceResult result = Template.exportOnce(
                templateFile(),
                Map.of(),
                Map.of(),
                ExportFormat.pdf(),
                CompilationMode.DEVELOPMENT
        );

        assertEquals("%PDF", new String(result.document(), 0, 4, StandardCharsets.US_ASCII));
        assertNull(result.warnings());
    }

    @Test
    void surfacesWarnings() throws IOException {
        byte[] template = packTemplate(
                MINIMAL_MANIFEST,
                "#set text(font: \"NonexistentFontExportOnce\")\nContent");

        ExportOnceResult result = Template.exportOnce(
                template,
                Map.of(),
                Map.of(),
                ExportFormat.svg(),
                CompilationMode.DEVELOPMENT
        );

        assertTrue(new String(result.document(), StandardCharsets.UTF_8).contains("<svg"));
        assertNotNull(result.warnings());
        assertTrue(result.warnings().contains("NonexistentFontExportOnce"));
    }

    @Test
    void enforcesZipLimits() throws IOException {
        byte[] bytes = templateFile();
        var exception = assertThrows(OicanaException.class, () -> Template.exportOnce(
                bytes,
                Map.of(),
                Map.of(),
                ExportFormat.pdf(),
                CompilationMode.DEVELOPMENT,
                null,
                new ZipLimits(1L, null)
        ));
        assertTrue(exception.getMessage().contains("entries"));
    }

    @Test
    void registrationEnforcesZipLimits() throws IOException {
        byte[] bytes = templateFile();
        var exception = assertThrows(OicanaException.class, () ->
                new Template(bytes, Map.of(), Map.of(), CompilationMode.DEVELOPMENT,
                        new ZipLimits(1L, null)));
        assertTrue(exception.getMessage().contains("entries"));
    }

    @Test
    void diagnosticColorConfigurationSucceeds() {
        Configuration.setDiagnosticColor(DiagnosticColor.ANSI);
        Configuration.setDiagnosticColor(DiagnosticColor.NONE);
    }
}
